Validation/HLTrigger/plugins/HLTGenValSource.cc

```diff
@@ -0,0 +1,477 @@
+  histConfigDefaults.push_back(histConfigDefault1);
+
+  desc.addVPSet("histConfigs", histConfig, histConfigDefaults);
+
+  // defining single histConfig2D
+  edm::ParameterSetDescription histConfig2D;
+  histConfig2D.add<std::string>("vsVarX");
+  histConfig2D.add<std::string>("vsVarY");
+  histConfig2D.add<std::vector<double>>("binLowEdgesX");
+  histConfig2D.add<std::vector<double>>("binLowEdgesY");
+
+  // default set of histConfigs
+  std::vector<edm::ParameterSet> histConfigDefaults2D;
+
+  edm::ParameterSet histConfigDefault2D0;
+  histConfigDefault2D0.addParameter<std::string>("vsVarX", "pt");
+  histConfigDefault2D0.addParameter<std::string>("vsVarY", "eta");
+  histConfigDefault2D0.addParameter<std::vector<double>>("binLowEdgesX", defaultPtBinning);
+  histConfigDefault2D0.addParameter<std::vector<double>>("binLowEdgesY", defaultetaBinning);
+  histConfigDefaults2D.push_back(histConfigDefault2D0);
+
+  desc.addVPSet("histConfigs2D", histConfig2D, histConfigDefaults2D);
+
+  // binnings, which are vectors of PSets
+  // there are no default for this
+  edm::ParameterSetDescription binningConfig;
+  binningConfig.add<std::string>("name");
+  binningConfig.add<std::string>("vsVar");
+  binningConfig.add<std::vector<double>>("binLowEdges");
+
+  // this by default is empty
+  std::vector<edm::ParameterSet> binningConfigDefaults;
+
+  desc.addVPSet("binnings", binningConfig, binningConfigDefaults);
+
+  descriptions.addDefault(desc);
+}
+
+// this method handles the different object types and collections that can be used for efficiency calculation
+std::vector<HLTGenValObject> HLTGenValSource::getObjectCollection(const edm::Event& iEvent) {
+  std::vector<HLTGenValObject> objects;  // the vector of objects to be filled
+
+  // handle object type
+  std::vector<std::string> implementedGenParticles = {"ele", "pho", "mu", "tau"};
+  if (std::find(implementedGenParticles.begin(), implementedGenParticles.end(), objType_) !=
+      implementedGenParticles.end()) {
+    objects = getGenParticles(iEvent);
+  } else if (objType_ == "AK4jet") {  // ak4 jets, using the ak4GenJets collection
+    const auto& genJets = iEvent.getHandle(ak4genJetToken_);
+    for (size_t i = 0; i < genJets->size(); i++) {
+      const reco::GenJet p = (*genJets)[i];
+      objects.emplace_back(p);
+    }
+  } else if (objType_ == "AK8jet") {  // ak8 jets, using the ak8GenJets collection
+    const auto& genJets = iEvent.getHandle(ak8genJetToken_);
+    for (size_t i = 0; i < genJets->size(); i++) {
+      const reco::GenJet p = (*genJets)[i];
+      objects.emplace_back(p);
+    }
+  } else if (objType_ == "AK4HT") {  // ak4-based HT, using the ak4GenJets collection
+    const auto& genJets = iEvent.getHandle(ak4genJetToken_);
+    if (!genJets->empty()) {
+      auto HTsum = 0.;
+      for (size_t i = 0; i < genJets->size(); i++) {
+        if (((*genJets)[i].pt() > 30) && (abs((*genJets)[i].eta()) < 2.5))
+          HTsum += (*genJets)[i].pt();
+      }
```

Comment:
This could be simplified:
```suggestion
      double HTsum = 0.;
      for (const auto& genjet : genJets) {
        if (genJet.pt() > 30 && std::abs(genJet.eta()) < 2.5)
          HTsum += genJet.pt();
      }
```